cardano-client-demo/MainPipeline.hs

```diff
@@ -0,0 +1,146 @@
+      ClientPipelinedStIdle(SendMsgDone, SendMsgRequestNextPipelined,
+                            CollectResponse),
+      ClientStNext(..) )
+import Network.TypedProtocol.Pipelined
+    ( N(..), Nat(..), natToInt, unsafeIntToNat )
+
+import Control.Monad (when)
+import Data.Kind
+import Data.Proxy
+import Data.Time
+import qualified GHC.TypeLits as GHC
+import System.Environment (getArgs)
+import System.FilePath ((</>))
+
+-- | Connects to a local cardano node, requests the blocks and prints out the
+-- number of transactions. To run this, you must first start a local node e.g.:
+--
+--     $ cabal run cardano-node:cardano-node -- run \
+--        --config        configuration/cardano/mainnet-config.json \
+--        --topology      configuration/cardano/mainnet-topology.json \
+--        --database-path db \
+--        --socket-path   db/node.sock \
+--        --host-addr     127.0.0.1 \
+--        --port          3001 \
+--
+-- Then run this with the path to the directory containing node.sock:
+--
+--     $ cabal run cardano-client-demo-pipeline -- db
+--
+main :: IO ()
+main = do
+  -- Get cocket path from CLI argument.
+  socketDir:_ <- getArgs
+  let socketPath = socketDir </> "node.sock"
+
+  -- Connect to the node.
+  putStrLn $ "Connecting to socket: " <> socketPath
+  IPC.connectToLocalNode
+    (connectInfo socketPath)
+    protocols
+  where
+  connectInfo :: FilePath -> IPC.LocalNodeConnectInfo IPC.CardanoMode
+  connectInfo socketPath =
+      IPC.LocalNodeConnectInfo {
+        IPC.localConsensusModeParams = IPC.CardanoModeParams (Byron.EpochSlots 21600),
+        IPC.localNodeNetworkId       = Mainnet,
+        IPC.localNodeSocketPath      = socketPath
+      }
+
+  protocols :: IPC.LocalNodeClientProtocolsInMode IPC.CardanoMode
+  protocols =
+      IPC.LocalNodeClientProtocols {
+        IPC.localChainSyncClient    = Just (Left (chainSyncClient (safeIntToNat (Proxy @50)))),
+        IPC.localTxSubmissionClient = Nothing,
+        IPC.localStateQueryClient   = Nothing
+      }
+
+-- What's the correct way to do this?
+safeIntToNat :: GHC.KnownNat ghcn => Proxy ghcn -> Nat (ToNat ghcn)
+safeIntToNat p = unsafeIntToNat (fromInteger $ GHC.natVal p)
+type family ToNat (n :: GHC.Nat) :: N where
+  ToNat 0 = Z
+  ToNat n = S (ToNat (n GHC.- 1))
+
+
+-- | Defines the pipelined client side of the chain sync protocol.
+chainSyncClient
+  :: Nat (S n)
+  -- ^ The maximum number of concurrent requests.
+  -> ChainSyncClientPipelined
+        (IPC.BlockInMode IPC.CardanoMode)
+        ChainPoint
+        ChainTip
+        IO
+        ()
+chainSyncClient pipelineSize = ChainSyncClientPipelined $ do
+  startTime <- getCurrentTime
+  let
+    clientIdle_RequestMoreN :: Nat n -> ClientPipelinedStIdle n (IPC.BlockInMode IPC.CardanoMode)
+                                 ChainPoint ChainTip IO ()
+    clientIdle_RequestMoreN n = case n of
+      Succ predN | natToInt n >= natToInt pipelineSize -> CollectResponse Nothing (clientNextN predN)
```

Comment:
Hmm. I think we should avoid using `natToInt` within the protocol, since it's horribly slow. The other uses of the type level nats is all O(1), to see if it's zero or not. I forget what exactly we did for pipeline depth policy in other cases (have a look) but I'm pretty sure we did not compare unary-rep typed nats.

This would also avoid needing `safeIntToNat` above. We don't need type level nats for the max pipelining. The typed nat is only needed to track the current pipeline depth so we know when it's safe to do certain things (in the case of zero outstanding replies).

Reply:
`natToInt` is actually just unwrapping a newtype:

```
natToInt :: Nat n -> Int
natToInt (UnsafeInt n) = n
```

It looks like we have `module Ouroboros.Network.Protocol.ChainSync.PipelineDecision` which helps to implement different pipelining strategies. I think I should use the `pipelineDecisionMax` strategy here.

Reply:
As per above, `natToInt` is great in terms of performance. Resolving.